yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryService.scala

```diff
@@ -0,0 +1,1048 @@
+   */
+  private[yarn] def registerListener(): Boolean = {
+    assert(sparkContext != null, "Null context")
+    if (sparkContext.conf.getBoolean(REGISTER_LISTENER, true)) {
+      logDebug("Registering listener to spark context")
+      val l = new YarnEventListener(sparkContext, this)
+      listener = Some(l)
+      sparkContext.listenerBus.addListener(l)
+      true
+    } else {
+      false
+    }
+  }
+
+  /**
+   * Queue an action, or if the service's `stopped` flag
+   * is set, discard it
+   * @param event event to process
+   * @return true if the event was queued
+   */
+  def enqueue(event: SparkListenerEvent): Boolean = {
+    if (!stopped.get) {
+      _eventsQueued.incrementAndGet
+      logDebug(s"Enqueue $event")
+      handleEvent(event)
+      true
+    } else {
+      if (timelineServiceEnabled) {
+        // if the timeline service was ever enabled, log the fact the event
+        // is being discarded. Don't do this if it was not, as it will
+        // only make the logs noisy.
+        logInfo(s"History service stopped; ignoring queued event : $event")
+      }
+      false
+    }
+  }
+
+  /**
+   * Stop the service; this triggers flushing the queue and, if not already processed,
+   * a pushing out of an application end event.
+   *
+   * This operation will block for up to `maxTimeToWaitOnShutdown` milliseconds
+   * to await the asynchronous action queue completing.
+   */
+  override def stop(): Unit = {
+    val oldState = enterState(StoppedState)
+    if (oldState != StartedState) {
+      // stopping from a different state
+      logDebug(s"Ignoring stop() request from state $oldState")
+      return
+    }
+    // if the queue is live
+    if (!stopped.get) {
+
+      if (appStartEventProcessed.get && !appEndEventProcessed.get) {
+        // push out an application stop event if none has been received
+        logDebug("Generating a SparkListenerApplicationEnd during service stop()")
+        enqueue(SparkListenerApplicationEnd(now()))
+      }
+
+      // flush out the events
+      asyncFlush()
+
+      // stop operation
+      postThreadActive.synchronized {
+        // now await that marker flag
+        if (postThreadActive.get) {
+          logDebug(s"Stopping posting thread and waiting $shutdownWaitTime mS")
+          stopped.set(true)
+          eventHandlingThread.foreach(_.interrupt())
+          postThreadActive.wait(shutdownWaitTime)
+        } else {
+          stopTimelineClient()
+          logInfo(s"Stopped: $this")
+        }
+      }
+    }
+  }
+
+  /**
+   * Can an event be added?
+   * Policy is only if the number of queued entities is below the limit, or the
+   * event marks the end of the application.
+   * @param isLifecycleEvent is this operation triggered by an application start/end?
+   * @return true if the event can be added to the queue
+   */
+  private def canAddEvent(isLifecycleEvent: Boolean): Boolean = {
+    _entityQueue.synchronized {
```

Comment:
Is the synchronization necessary here? The queue itself is thread-safe, and you're only adding to the queue from the listener bus thread from what I understand. So you won't have two writers trying to add events, which would actually cause a race in `handleEvent()`.